File: fe1-web/src/features/social/network/messages/chirp/DeleteChirp.ts
import { ActionType, MessageData, ObjectType } from 'core/network/jsonrpc/messages';
import { validateDataObject } from 'core/network/validation';
import { checkTimestampStaleness } from 'core/network/validation/Checker';
import { Hash, ProtocolError, Timestamp } from 'core/objects';

/**
 * Data sent to remove a chirp
 */
export class DeleteChirp implements MessageData {
  public readonly object: ObjectType = ObjectType.CHIRP;

  public readonly action: ActionType = ActionType.DELETE;

  // The id of the chirp published
  public readonly chirp_id: Hash;

  // The timestamp of this deletion request
  public readonly timestamp: Timestamp;

  constructor(msg: Partial<DeleteChirp>) {
    if (!msg.chirp_id) {
      throw new ProtocolError("Undefined 'chirp_id' parameter encountered during 'RemoveChirp'");
    }
    this.chirp_id = msg.chirp_id;

    if (!msg.timestamp) {
      throw new ProtocolError("Undefined 'timestamp' parameter encountered during 'RemoveChirp'");
    }
    checkTimestampStaleness(msg.timestamp);
    this.timestamp = msg.timestamp;
  }

  /**
   * Creates a RemoveChirp object from a given object
   * @param obj
   */
  public static fromJson(obj: any): DeleteChirp {
    const { errors } = validateDataObject(ObjectType.CHIRP, ActionType.DELETE, obj);

    if (errors !== null) {
      throw new ProtocolError(`Invalid chirp remove\n\n${errors}`);
    }

    return new DeleteChirp({
      chirp_id: new Hash(obj.chirp_id),
      timestamp: new Timestamp(obj.timestamp),
    });
  }
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/socialMedia/DeleteChirp.scala
package ch.epfl.pop.model.network.method.message.data.socialMedia

import ch.epfl.pop.json.MessageDataProtocol._
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.ActionType.ActionType
import ch.epfl.pop.model.network.method.message.data.ObjectType.ObjectType
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Hash, Timestamp}
import spray.json._

final case class DeleteChirp(
    chirp_id: Hash,
    timestamp: Timestamp
) extends MessageData {
  override val _object: ObjectType = ObjectType.CHIRP
  override val action: ActionType = ActionType.DELETE
}

object DeleteChirp extends Parsable {
  def apply(
      chirp_id: Hash,
      timestamp: Timestamp
  ): DeleteChirp = new DeleteChirp(chirp_id, timestamp)

  override def buildFromJson(payload: String): DeleteChirp = payload.parseJson.asJsObject.convertTo[DeleteChirp]
}
